plugins/wallabag.koplugin/main.lua

```diff
@@ -622,270 +692,457 @@ function Wallabag:download(article)
+-- @param[opt=false] quiet
+-- @treturn bool Whether the request was successful
+-- @treturn string Error type if unsuccessful, filepath if success with path, JSON if without
+-- @treturn int HTTP response code if unsuccessful (e.g. 404, 503, …)
+function Wallabag:callAPI(method, url, headers, body, filepath, quiet)
+    quiet = quiet or false
+
     local sink = {}
-    local request = {}
+    local request = { method = method }
 
     -- Is it an API call, or a regular file direct download?
-    if apiurl:sub(1, 1) == "/" then
+    --- @todo Separate call to internal API from the download on external server
+    if url:sub(1, 1) == "/" then
         -- API call to our server, has the form "/random/api/call"
-        request.url = self.server_url .. apiurl
-        if headers == nil then
-            headers = {
-                ["Authorization"] = "Bearer " .. self.access_token,
-            }
-        end
-    else
-        -- regular url link to a foreign server
-        local file_url = apiurl
-        request.url = file_url
-        if headers == nil then
-            -- no need for a token here
-            headers = {}
-        end
+        request.url = self.server_url .. url
+        request.headers = headers or {
+            ["Authorization"] = "Bearer " .. self.access_token,
+        }
+    else -- Assume full URL (e.g. https://…)
+        request.url = url
+        request.headers = headers or {}
     end
 
-    request.method = method
-    if filepath ~= "" then
+    if filepath ~= nil then
         request.sink = ltn12.sink.file(io.open(filepath, "w"))
         socketutil:set_timeout(socketutil.FILE_BLOCK_TIMEOUT, socketutil.FILE_TOTAL_TIMEOUT)
     else
         request.sink = ltn12.sink.table(sink)
         socketutil:set_timeout(socketutil.LARGE_BLOCK_TIMEOUT, socketutil.LARGE_TOTAL_TIMEOUT)
     end
-    request.headers = headers
-    if body ~= "" then
+
+    if body ~= nil then
         request.source = ltn12.source.string(body)
     end
-    logger.dbg("Wallabag: URL     ", request.url)
-    logger.dbg("Wallabag: method  ", method)
+
+    logger.dbg("Wallabag:callAPI:", request.method, request.url)
 
     local code, resp_headers, status = socket.skip(1, http.request(request))
     socketutil:reset_timeout()
-    -- raise error message when network is unavailable
+
+    -- Raise error if network is unavailable
     if resp_headers == nil then
-        logger.dbg("Wallabag: Server error:", status or code)
-        self:removeFailedDownload(filepath)
-        return nil, "network_error"
+        if filepath then
+            self:removeFailedDownload(filepath)
+        end
+
+        logger.err("Wallabag:callAPI: network error", status or code)
+        return false, "network_error"
     end
+
+    -- If the request returned successfully
     if code == 200 then
-        if filepath ~= "" then
-            logger.dbg("Wallabag: file downloaded to", filepath)
-            return true
+        if filepath then
```

Comment:
Same suspicious change.
```suggestion
        if filepath then
```